core/shared/src/main/scala/fs2/Chunk.scala

```diff
@@ -945,19 +959,36 @@ object Chunk
       else new Queue(chunks :+ that, size + that.size)
 
     /** Prepends a chunk to the start of this chunk queue. */
-    def +:[O2 >: O](c: Chunk[O2]): Queue[O2] = new Queue(c +: chunks, c.size + size)
+    def +:[O2 >: O](c: Chunk[O2]): Queue[O2] =
+      if (c.isEmpty) this else new Queue(c +: chunks, c.size + size)
 
     /** Appends a chunk to the end of this chunk queue. */
-    def :+[O2 >: O](c: Chunk[O2]): Queue[O2] = new Queue(chunks :+ c, size + c.size)
+    def :+[O2 >: O](c: Chunk[O2]): Queue[O2] =
+      if (c.isEmpty) this else new Queue(chunks :+ c, size + c.size)
 
     def apply(i: Int): O = {
       if (i < 0 || i >= size) throw new IndexOutOfBoundsException()
-      def go(chunks: SQueue[Chunk[O]], offset: Int): O = {
-        val head = chunks.head
-        if (offset < head.size) head(offset)
-        else go(chunks.tail, offset - head.size)
+      if (i == 0) chunks.head(0)
+      else if (i == size - 1) chunks.last.last.get
+      else {
+        val (lengths, chunks) = accumulatedLengths
+        var lo = 0
+        var hi = lengths.length - 1
+        var mid = (hi - lo) / 2
+        var accLen = 0
+        var chunk: Chunk[O] = null
+        while (lo <= hi) {
+          val delta = hi - lo
+          mid = lo + (delta / 2)
+          accLen = lengths(mid)
+          chunk = chunks(mid)
+          if (i >= accLen) lo = if (delta <= 1) mid + 1 else mid
+          else if (i >= accLen - chunk.size) lo = hi + 1
+          else hi = mid
+        }
+        val accLenBefore = accLen - chunk.size
+        chunk(i - accLenBefore)
```

Comment:
```suggestion
        // 1. What is the chunk that would contain the element? 
        var binS = Arrays.binarySearch(lengths, i)
        if (binS >= 0) {
          /* Index `i` is contained in lenghts array, which means one sum of 
          * chunks lengths is in array, so `i` is head of next chunk 
            Note that, since `i < size`, binS must be less than `lengths.length - 1` */
          chunk(binS + 1).head 
        } else {
          /* binS is (- insertion point - 1), so the index `i` is at the chunk at insertion point */ 
          val j = - binS + 1 
          val accLenBefore = if (j == 0) 0 else lengths(j-1)
          val chunk = chunks(j)
          chunk(i - accLenBefore)
        }
```
I think this should work, but given that this is array indexitis, I would not send it forth without a few tests first. 

Note that this is taking into account that we start our `lengths` with the length of the first chunk, and it ends with the size of the whole queue.

Reply:
Oh cool, I didn't realize the Java built-in binary search returned the negative insertion point. That'll definitely work.

What's your concern regarding arrays? A Scala array *is* a Java array.